pkg/vault/contracts/CompositeLiquidityRouter.sol

```diff
@@ -0,0 +1,512 @@
+                userData: params.userData
+            })
+        );
+
+        for (uint256 i = 0; i < parentPoolTokens.length; i++) {
+            address childPool = address(parentPoolTokens[i]);
+
+            if (_vault.isPoolRegistered(childPool)) {
+                // Token is a BPT, so remove liquidity from the child pool.
+
+                // We don't expect the sender to have BPT to burn. So, we flashloan tokens here (which should in
+                // practice just use existing credit).
+                _vault.sendTo(IERC20(childPool), address(this), parentPoolAmountsOut[i]);
+
+                IERC20[] memory childPoolTokens = _vault.getPoolTokens(childPool);
+                // Router is an intermediary in this case. The Vault will burn tokens from the router, so Router is
+                // both owner and spender (which doesn't need approval).
+                (, uint256[] memory childPoolAmountsOut, ) = _vault.removeLiquidity(
+                    RemoveLiquidityParams({
+                        pool: childPool,
+                        from: address(this),
+                        maxBptAmountIn: parentPoolAmountsOut[i],
+                        minAmountsOut: new uint256[](childPoolTokens.length),
+                        kind: params.kind,
+                        userData: params.userData
+                    })
+                );
+                // Return amounts to user.
+                for (uint256 j = 0; j < childPoolTokens.length; j++) {
+                    _currentSwapTokensOut().add(address(childPoolTokens[j]));
+                    _currentSwapTokenOutAmounts().tAdd(address(childPoolTokens[j]), childPoolAmountsOut[j]);
+                }
+            } else {
+                // Token is not a BPT, so return the amount to the user.
+                _currentSwapTokensOut().add(childPool);
+                _currentSwapTokenOutAmounts().tAdd(childPool, parentPoolAmountsOut[i]);
+            }
+        }
+
+        // The hook writes current swap token and token amounts out.
+        // We copy that information to memory to return it before it is deleted during settlement.
+        tokensOut = InputHelpers.sortTokens(_currentSwapTokensOut().values().asIERC20()).asAddress();
+        amountsOut = new uint256[](tokensOut.length);
+        for (uint256 i = 0; i < tokensOut.length; ++i) {
+            amountsOut[i] = _currentSwapTokenOutAmounts().tGet(tokensOut[i]);
+
+            if (amountsOut[i] < params.minAmountsOut[i]) {
+                revert IVaultErrors.AmountOutBelowMin(IERC20(tokensOut[i]), amountsOut[i], params.minAmountsOut[i]);
+            }
+        }
+
+        _settlePaths(params.sender, false);
+    }
+
+    /*******************************************************************************
+                                    Settlement
+    *******************************************************************************/
+
+    function _settlePaths(address sender, bool wethIsEth) internal {
```

Comment:
This is the same as with batch router right? Technically there are no 'paths' here.
We don't have a common layer between batch and composite router though :/ 

Do you think there's a good way of sharing this code between the routers?

Reply:
I thought about adding it to RouterCommon, but I wasn't sure since the Router doesn't use it. Anyway, I can add it in there.